### PR TITLE
Proof-of-concept Dashboard

### PR DESCRIPTION
**Pull request by dawiddawidowski**
A proof of concept multi-page dashboard.

The dashboard contains the following features:
- Home page. 
  - Includes a welcome message and guides the user to select one of the pages in the sidebar.
 
- Current rides page:
  - Header containing current rider's full name
  - Sub-header indicating the time since the data was last updated 
  - Expandable `Personal info` header for user convenience
  - Permanent metrics header containing the  elapsed time, heart rate, power and resistance
  - Heart rate warning header if the heart rate exceeds the safety threshold based on the current rider's gender and age.
  - Expandable `Personal best` header for user convenience, including the same metrics as the metrics header.
  
- Recent rides page (for the past 12 hours of data):
  - Displays a row of the widget containing bar charts for the total elapsed time per gender, the total number of rides per gender and the total number of rides per age bracket (Under 18, 18-24, 25-34, 35-44, 45-54, 55-64, 65+)
  - Displays a row of the widget containing line charts for the average power output and average resistance over the past 12 hours.
  - Displays a row of the widget containing line charts for the cumulative power output and cumulative resistance over the past 12 hours.
  - Charts are interactive, enabling filtering of data in all graphs by gender by clicking on a bar of the gender-categorised bar charts.
  - Side bar includes a slider to allow the user to choose the resolution of the time in the line charts.
  

- Restructured the folder structure:
  - `dashboard.py` contains the code for the dashboard home page
  - `pages` folder containing the pages of the dashboard:
    - `1_Current_Rides.py` - contains the code for the current rides (same state as the before used main file `dashboard.py`)
    - `2_Recent_Rides.py` - contains the code for the recent rides over the past 12 hours.

- `utilities.py` - added functions to process the dataframe containing the columns necessary to produce the visualisations.

- `visualisations.py` - added functions to generate the visualisations for the recent rides page

- `database.py` - added functions to send SQL queries to the RDS database for rides data in the past 12 hours.

### CODE REVIEW

File: dashboard/dashboard.py [hunk 1]
```diff
@@ -1,74 +1,34 @@
 """
-Dashboard script to establish connection to the RDS database, fetch data using SQL queries and
-create visualisations in a Streamlit app (using functions from the `database.py`,
-`visualisations.py` and `utilities.pyz files as necessary).
+Entry point home page for the dashboard.
```
Comment (CharliedeanCS) [inline dashboard/dashboard.py:2]: Can all this code not be in a if __name__ == __main__?

File: dashboard/database.py [hunk 2]
```diff
@@ -22,18 +29,27 @@ def get_database_connection() -> extensions.connection:
...
+        )
+        SELECT
+            current_ride.rider_id, first_name, last_name, height, weight, gender, birthdate,
+            heart_rate, power, resistance, elapsed_time, start_time
+        FROM current_ride
+        JOIN Rider ON current_ride.rider_id = Rider.rider_id
+        JOIN Reading ON current_ride.ride_id = Reading.ride_id
         ORDER BY start_time DESC, elapsed_time DESC
         LIMIT 1;
         """
```
Comment (CharliedeanCS) [inline dashboard/database.py:55]: Nicely done with the query

File: dashboard/database.py [hunk 3]
```diff
@@ -155,3 +171,107 @@ def get_current_ride_data_highest(db_connection: extensions.connection, rider_de
...
+
+
+# RECENT RIDES
+def get_recent_12hr_data(db_connection: extensions.connection) -> pd.DataFrame:
+    """
+    Retrieves data from the last 12 hours (by attribute 'start_time') from the database
+    as a Pandas Dataframe.
+    """
+
+    twelve_hours_ago = (datetime.now() - timedelta(hours=12))
```
Comment (CharliedeanCS) [inline dashboard/database.py:183]: Why 12 hours? Why not the previous days data?

File: dashboard/dashboard.py (hunk 1 above)
Comment (dawiddawidowski) [reply dashboard/dashboard.py:2]: Unfortunately no :/ (link if you want to check out the Streamlit tutorial https://docs.streamlit.io/get-started/tutorials/create-a-multipage-app)

File: dashboard/database.py [hunk 4]
```diff
@@ -155,3 +171,107 @@ def get_current_ride_data_highest(db_connection: extensions.connection, rider_de
...
+
+        db_cur.execute(query, parameters)
+
+        recent_rides = db_cur.fetchall()
+
+        columns = [
+            "rider_id", "first_name", "last_name", "height", "weight",
+            "gender", "birthdate", "heart_rate", "power",
+            "resistance", "elapsed_time", "start_time", "ride_id"
+        ]
```
Comment (CharliedeanCS) [inline dashboard/database.py:207]: Nice and understandable names

File: dashboard/database.py [hunk 5]
```diff
@@ -155,3 +171,107 @@ def get_current_ride_data_highest(db_connection: extensions.connection, rider_de
...
+
+        return pd.DataFrame(recent_rides, columns=columns)
+
+
+def get_ride_count_gender(db_connection: extensions.connection) -> pd.DataFrame:
+    """
+    Retrieves data from the last 12 hours (by attribute 'start_time') from the database
+    as a Pandas Dataframe.
+    """
```
Comment (CharliedeanCS) [inline dashboard/database.py:216]: Nice docstrings throughout

File: dashboard/database.py [hunk 6]
```diff
@@ -155,3 +171,107 @@ def get_current_ride_data_highest(db_connection: extensions.connection, rider_de
...
+        """
+
+        parameters = (twelve_hours_ago, )
+
+        db_cur.execute(query, parameters)
+
+        response = db_cur.fetchall()[0]
+
+        age_brackets = ["0-18", "18-24", "25-34",
+                        "35-44", "45-54", "55-64", "65+"]
```
Comment (CharliedeanCS) [inline dashboard/database.py:272]: How did you decide the age bracket?

File: dashboard/pages/1_Current_Ride.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
+"""
+Dashboard script for the current ride to establish connection to the RDS database,
+fetch data using SQL queries and create visualisations in a Streamlit app
+using functions from the `database.py`, `visualisations.py` and `utilities.pyz files as necessary.
+"""
```
Comment (CharliedeanCS) [inline dashboard/pages/1_Current_Ride.py:5]: Nice and descriptive module docstring

File: dashboard/pages/2_Recent_Rides.py [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+        recent_rides, selector_gender)
+
+    cumul_resistance_chart = get_resistance_output_cumul_line_chart(
+        recent_rides, selector_gender)
+
+    # try to add empty bar joining to space horizontally
+
+    row1 = avg_power_chart | avg_resistance_chart
+    row2 = cumul_power_chart | cumul_resistance_chart
+    line_graphs = row1 & row2
```
Comment (CharliedeanCS) [inline dashboard/pages/2_Recent_Rides.py:71]: Again row1 and row2 could have better names.

File: dashboard/utilities.py [hunk 9]
```diff
@@ -30,12 +35,15 @@ def calculate_max_heart_rate(user_details: list) -> int:
...
+
+    'other' and 'None' gender heart rates are treated conservatively using the formula
+    for females as a safety precaution.
     """
     birthdate = user_details[6]
     age = calculate_age(birthdate)
     gender = user_details[5]
 
-    if gender == "female":
+    if gender in ("female", "other", None):
```
Comment (CharliedeanCS) [inline dashboard/utilities.py:46]: Again, this is fine. We may have to justify why we decided to put other and None in this bracket

File: dashboard/utilities.py [hunk 10]
```diff
@@ -78,7 +89,124 @@ def is_heart_rate_abnormal(user_details: list) -> bool:
...
+    df = recent_rides[['elapsed_time',
+                       'start_time', 'power', 'resistance']].copy()
+
+    delta = timedelta(minutes=date_resolution)
+
+    df["reading_time"] = df.apply(
+        lambda x: (
+            pd.to_datetime(x['start_time']) +
+            pd.to_timedelta(x['elapsed_time'], unit='s')
+        ).round('min'), axis=1)
```
Comment (CharliedeanCS) [inline dashboard/utilities.py:155]: Very nice use of a lambda

File: dashboard/utilities.py [hunk 11]
```diff
@@ -78,7 +89,124 @@ def is_heart_rate_abnormal(user_details: list) -> bool:
...
+    """Returns a DataFrame that average power output per minute."""
+
+    grouped_df = ride_data.groupby(
+        ['reading_time'])['power'].mean().astype(int)
+
+    grouped_df = grouped_df.reset_index()
+
+    grouped_df.columns = ['reading_time', 'power']
+
+    return grouped_df
```
Comment (CharliedeanCS) [inline dashboard/utilities.py:173]: grouped_df could have a better name. What is it grouped by?

File: dashboard/visualisations.py [hunk 12]
```diff
@@ -95,9 +97,142 @@ def get_current_ride_metrics(user_details: list) -> None:
...
+def get_recent_rides_header() -> None:
+    """Generates a header for the recent rides section."""
+    st.header("RECENT RIDES", divider='green')
+
+
+def get_last_updated_recent_rides(last_update_time: datetime,
+                                  _last_updated_placeholder: st.empty) -> None:
+    """Generates a caption under the header with the time since the last data update."""
+
+    # time_delta = (current_time-last_update_time).total_seconds()
```
Comment (CharliedeanCS) [inline dashboard/visualisations.py:120]: Can this be deleted?

**Review by CharliedeanCS — APPROVED** [review]
Dashboard looks nice. Code is high quality. Everything works correctly and is tested

**Pull request merged**